src/ImageSharp/Processing/Processors/Quantization/EuclideanPixelMap{TPixel}.cs

```diff
@@ -97,9 +97,92 @@ private int GetClosestColorSlow(TPixel color, ref TPixel paletteRef, out TPixel
             }
 
             // Now I have the index, pop it into the cache for next time
-            this.distanceCache[color] = index;
+            this.cache.Add(rgba, (byte)index);
             match = Unsafe.Add(ref paletteRef, index);
             return index;
         }
+
+        /// <summary>
+        /// Returns the Euclidean distance squared between two specified points.
+        /// </summary>
+        /// <param name="a">The first point.</param>
+        /// <param name="b">The second point.</param>
+        /// <returns>The distance squared.</returns>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        private static float DistanceSquared(Rgba32 a, Rgba32 b)
+        {
+            int deltaR = a.R - b.R;
+            int deltaG = a.G - b.G;
+            int deltaB = a.B - b.B;
+            int deltaA = a.A - b.A;
+            return (deltaR * deltaR) + (deltaG * deltaG) + (deltaB * deltaB) + (deltaA * deltaA);
+        }
+
+        /// <summary>
+        /// A cache for storing color distance matching results.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// The granularity of the cache has been determined based upon the current
+        /// suite of test images and provides the lowest possible memory usage while
+        /// providing enough match accuracy.
+        /// Entry count is currently limited to 1221858 entries at 1.17MB.
+        /// </para>
+        /// </remarks>
+        private struct ColorDistanceCache
+        {
+            private const int IndexBits = 5;
+            private const int IndexAlphaBits = 4;
+            private const int IndexCount = (1 << IndexBits) + 1;
+            private const int IndexAlphaCount = (1 << IndexAlphaBits) + 1;
+            private const int RgbShift = 8 - IndexBits;
+            private const int AlphaShift = 8 - IndexAlphaBits;
+            private const int TableLength = IndexCount * IndexCount * IndexCount * IndexAlphaCount;
+            private short[] table;
+
+            public static ColorDistanceCache Create()
+            {
+                ColorDistanceCache result = default;
+                short[] entries = new short[TableLength];
```

Comment:
We need to use `MemoryAllocator` here. Fortunately `IQuantizer<TPixel> : IDisposable`, so it shouldn't be an issue.

Reply:
Although realized we need fast access through `Span<short>` instead of `Memory<short>`, which seems to be impossible with the current design.

We can keep the LOH allocation or pin the pooled buffer like in `ResizeKernelMap`:
https://github.com/SixLabors/ImageSharp/blob/381dff8640b721a34b1227c970fcf6ad6c5e3e72/src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeKernelMap.cs#L55
https://github.com/SixLabors/ImageSharp/blob/381dff8640b721a34b1227c970fcf6ad6c5e3e72/src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeKernelMap.cs#L258

Reply:
Should be doable. I’ll have a look

Reply:
Pinning a 1MB buffer is not ideal, but if it happens for a <100ms time, it's probably better than a LOH allocation of the same amount.

Reply:
Is there a risk using `MemoryAllocator` that we could end up with a discontiguous memory buffer.

Reply:
Only if you use Allocate2D

Reply:
@antonfirsov isn't 1mb array a LOH allocation? Most likely it won't be moved by GC anyway, only if user manually force `GCSettings.LargeObjectHeapCompactionMode`. So pinning shouldn't change anything, doesn't it? If memory pool gives native memory, it isn't moved either, no?

Reply:
Yeah thats true, forgot that LOH objects are not moved. Pinning wont hurt at all here.

Reply:
The pinning is more a performance thing to workaround the API to prevent us having to call `IMemoryOwner.GetSpan()` on a per pixel basis.